components/tikv_util/src/config.rs

```diff
@@ -457,17 +457,63 @@ impl<'de> Deserialize<'de> for ReadableDuration {
     }
 }
 
+fn canonicalize_imp<P: AsRef<Path>>(path: P) -> std::io::Result<PathBuf> {
+    fn canonicalize_non_existing_path<P: AsRef<Path>>(path: P) -> std::io::Result<PathBuf> {
+        use std::path::Component;
+        let mut components = path.as_ref().components().peekable();
+        let mut last_canonicalized = None;
+        let mut should_canonicalize = true;
+        let mut ret = if let Some(c @ Component::Prefix(..)) = components.peek().cloned() {
+            components.next();
+            PathBuf::from(c.as_os_str())
+        } else {
+            PathBuf::new()
+        };
+        for component in components {
+            match component {
+                Component::Prefix(..) => unreachable!(),
+                Component::RootDir => {
+                    ret.push(component.as_os_str());
+                }
+                Component::CurDir => {}
+                Component::ParentDir => {
+                    ret.pop();
```

Comment:
It is possible that we could still canonicalize after we pop the parent directory.

e.g.

```
realpath /folder1/folder2/non_exist/x/../../folder_3
```

where
```
/folder1 -> /folder2
/folder2/folder_3 -> /folder_4
```

should be resolved to `/folder_4`.

However, as `/folder1/folder2/non_exist` does not exist, this function won't do further canonicalize on `/folder1/folder2/folder_3`, if I understood this part correctly.